robot-server/robot_server/persistence.py

```diff
@@ -0,0 +1,196 @@
+_protocol_directory = AppStateValue[Path]("protocol_directory")
+
+_TEMP_PERSISTENCE_DIR_PREFIX: Final = "opentrons-robot-server-"
+_DATABASE_FILE: Final = "robot_server.db"
+_PROTOCOL_FILES_SUBDIRECTORY: Final = "protocols"
+
+_log = logging.getLogger(__name__)
+
+_metadata = sqlalchemy.MetaData()
+
+protocol_table = sqlalchemy.Table(
+    "protocol",
+    _metadata,
+    sqlalchemy.Column(
+        "id",
+        sqlalchemy.String,
+        primary_key=True,
+    ),
+    sqlalchemy.Column(
+        "created_at",
+        sqlalchemy.DateTime,
+        nullable=False,
+    ),
+    # TODO(mm, 2022-03-29):
+    # Storing pickled Python objects, especially of an internal class,
+    # will cause migration and compatibility problems.
+    sqlalchemy.Column(
+        "source",
+        sqlalchemy.PickleType,
+        nullable=False,
+    ),
+    sqlalchemy.Column("protocol_key", sqlalchemy.String, nullable=True),
+)
+
+run_table = sqlalchemy.Table(
+    "run",
+    _metadata,
+    sqlalchemy.Column(
+        "id",
+        sqlalchemy.String,
+        primary_key=True,
+    ),
+    sqlalchemy.Column(
+        "created_at",
+        sqlalchemy.DateTime,
+        nullable=False,
+    ),
+    sqlalchemy.Column(
+        "protocol_id",
+        sqlalchemy.String,
+        # TODO (tz 4/8/22): SQLite does not support FK by default. Need to add support
+        # https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#foreign-key-support
+        sqlalchemy.ForeignKey("protocol.id"),
+        nullable=True,
+    ),
+)
+
+# TODO (tz: 4/8/22): add a column sequence_number for preserving the order of actions
+actions_table = sqlalchemy.Table(
+    "action",
+    _metadata,
+    sqlalchemy.Column(
+        "id",
+        sqlalchemy.String,
+        primary_key=True,
+    ),
+    sqlalchemy.Column("created_at", sqlalchemy.DateTime, nullable=False),
+    sqlalchemy.Column("action_type", sqlalchemy.String, nullable=False),
+    sqlalchemy.Column(
+        "run_id",
+        sqlalchemy.String,
+        # TODO (tz 4/8/22): SQLite does not support FK by default. Need to add support
+        # https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#foreign-key-support
+        sqlalchemy.ForeignKey("run.id"),
+    ),
+)
+
+
+# TODO(mm, 2022-03-29): When we confirm we can use SQLAlchemy 1.4 on the OT-2,
+# convert these to return an async engine.
+# https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html
+@contextmanager
+def opened_db(db_file_path: Path) -> Generator[SQLEngine, None, None]:
+    """Return an Engine for a SQLite database on hte filesystem.
+
+    Clean up the engine when the context manager exits.
+
+    The new database will be totally blank--no tables.
+    """
+    sql_engine = open_db_no_cleanup(db_file_path=db_file_path)
+    try:
+        yield sql_engine
+    finally:
+        sql_engine.dispose()
+
+
+def open_db_no_cleanup(db_file_path: Path) -> SQLEngine:
+    """Like `create_in_memory_db()`, except without automatic cleanup."""
```

Comment:
I think this docstring is out of date. Is this what it's trying to say?

```suggestion
    """Create a database engine for performing transactions."""
```